controllers/argocd/role.go

```diff
@@ -343,3 +372,141 @@ func deleteClusterRoles(c client.Client, clusterRoleList *v1.ClusterRoleList) er
 	}
 	return nil
 }
+
+// checkCustomClusterRoleMode checks if custom ClusterRole mode is enabled and deletes default ClusterRoles if they exist
+func checkCustomClusterRoleMode(r *ReconcileArgoCD, cr *argoproj.ArgoCD, componentName string, allowed bool) (bool, error) {
+	// check if it is cluster-scoped instance namespace and user doesn't want to use default ClusterRole
+	if allowed && cr.Spec.DefaultClusterScopedRoleDisabled {
+
+		// in case DefaultClusterScopedRoleDisabled was false earlier and default ClusterRole was created, then delete it.
+		existingClusterRole := &v1.ClusterRole{}
+		if err := r.Client.Get(context.TODO(), types.NamespacedName{Name: GenerateUniqueResourceName(componentName, cr)}, existingClusterRole); err == nil {
+
+			// default ClusterRole exists, now delete it
+			if err := r.Client.Delete(context.TODO(), existingClusterRole); err != nil {
+				return true, fmt.Errorf("failed to delete existing cluster role for the service account associated with %s : %s", componentName, err)
+			}
+		}
+
+		// don't create a default ClusterRole and return
+		return true, nil
```

Comment:
In this function we are assuming that any error is a 'resource not found' error, which is not necessarily true.

We should explicitly check that the err is a not found error using `apierrors.IsNotFound`. For any other errors, we should return them.